### PR TITLE
Bounded params

### PR DESCRIPTION
**Pull request by shaunxw**
Introduce abstract `BoundedType` and `FractionalRate`. Will add other bounded types in follow up PRs.

### CODE REVIEW

File: modules/support/src/bounded.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		let (min, max) = Range::get();
+		if value < min || value > max {
+			return Err(Error::OutOfBound);
+		}
+		Ok(Self(value, PhantomData))
+	}
+
+	/// Set the inner value. Returns `Err` if out of bound or the diff with current value exceeds
+	/// the max absolute value.
+	pub fn set(&mut self, value: T) -> Result<(), Error> {
```
Comment (xlc) [inline modules/support/src/bounded.rs:77]: `try_set`?

File: modules/cdp-engine/src/lib.rs [hunk 2]
```diff
@@ -488,11 +493,21 @@ pub mod module {
 			T::UpdateOrigin::ensure_origin(origin)?;
 
 			let mut collateral_params = Self::collateral_params(currency_id).unwrap_or_default();
-			if let Change::NewValue(update) = interest_rate_per_sec {
-				collateral_params.interest_rate_per_sec = update;
+			if let Change::NewValue(maybe_rate) = interest_rate_per_sec {
+				if let (Some(existing), Some(rate)) = (collateral_params.interest_rate_per_sec.as_mut(), maybe_rate) {
```
Comment (xlc) [inline modules/cdp-engine/src/lib.rs:497]: better to make it a match

File: modules/support/src/bounded.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+impl<T: Default + Encode + Decode, Range: Get<(T, T)>, MaxChangeAbs: Get<T>> Default
+	for BoundedType<T, Range, MaxChangeAbs>
+{
+	fn default() -> Self {
+		let (min, _) = Range::get();
+		Self(min, PhantomData)
+	}
+}
+
+impl<T: Encode + Decode + CheckedSub + PartialOrd + Copy, Range: Get<(T, T)>, MaxChangeAbs: Get<T>>
```
Comment (xlc) [inline modules/support/src/bounded.rs:63]: should implement `TryFrom` and `Into` trait

Comment (shaunxw) [reply modules/support/src/bounded.rs:63]: `TryFrom` and `Into` cannot be implemented for bounded types, which conflict with impl in `core`, due to rust trait limitations https://github.com/rust-lang/rust/issues/50133

File: modules/cdp-engine/src/lib.rs [hunk 4]
```diff
@@ -87,7 +88,7 @@ pub struct RiskManagementParams {
 	pub maximum_total_debit_value: Balance,
 
 	/// Extra interest rate per sec, `None` value means not set
-	pub interest_rate_per_sec: Option<Rate>,
+	pub interest_rate_per_sec: Option<FractionalRate>,
```
Comment (wangjj9219) [inline modules/cdp-engine/src/lib.rs:91]: the bound and change limit check of `FractionalRate` doesn't make much sense for the `interest_rate_per_sec` of `RiskManagementParams`,  the config value is too tiny(for example 0.000000000860244)

Comment (shaunxw) [reply modules/cdp-engine/src/lib.rs:91]: We can definitely further limit the range, if reasonable bounds can be defined. Any suggestion for the upper limit of `interest_rate_per_sec`?

File: modules/support/src/bounded.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		let abs = if value > *old_value {
+			value
+				.checked_sub(old_value)
+				.expect("greater number subtracting smaller one can't underflow; qed")
+		} else {
+			old_value
+				.checked_sub(&value)
+				.expect("greater number subtracting smaller one can't underflow; qed")
+		};
+		if abs > max_change_abs {
```
Comment (wangjj9219) [inline modules/support/src/bounded.rs:103]: we may want to directly set the value to min for some situations, but if the change exceeds the  max_change_abs when try_set min, whether to bypass this check or not? @xlc

Comment (shaunxw) [reply modules/support/src/bounded.rs:103]: In this case you can make multiple calls in a batch, to gradually set the value until min is reached.

File: modules/cdp-engine/src/lib.rs (hunk 4 above)
Comment (wangjj9219) [reply modules/cdp-engine/src/lib.rs:91]: we can define reasonable bound for system params in #2362

**Review by xlc — APPROVED** [review]

**Pull request merged**